File: modeler/src/main/java/net/sayaya/document/modeler/sample/SampleRouter.java
package net.sayaya.document.modeler.sample;

import net.sayaya.document.data.Sample;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.codec.multipart.FilePart;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import static org.springframework.web.reactive.function.server.RequestPredicates.*;
import static org.springframework.web.reactive.function.server.RouterFunctions.route;

@Configuration
public class SampleRouter {
	private final SampleHandler handler;
	public SampleRouter(SampleHandler handler) {this.handler = handler;}
	@Bean
	public RouterFunction<ServerResponse> documentRouterInstance() {
		return route(GET("/models/{model}/samples"), this::findSamples)
				.andRoute(POST("/models/{model}/samples"), this::uploadDocuments);
	}
	private Mono<ServerResponse> findSamples(ServerRequest request) {
		String model = request.pathVariable("model");
		return handler.list(model).collectList()
				.flatMap(list->ServerResponse.ok().contentType(MediaType.APPLICATION_JSON).bodyValue(list))
				.switchIfEmpty(ServerResponse.noContent().build())
				.onErrorResume(e->ServerResponse.badRequest().bodyValue(e.getMessage()));
	}
	private Mono<ServerResponse> uploadDocuments(ServerRequest request) {
		String model = request.pathVariable("model");
		Flux<Sample> result = handler.upload(model, request.multipartData().map(i->i.get("files")).flatMapMany(Flux::fromIterable).cast(FilePart.class));
		return ServerResponse.ok().contentType(MediaType.APPLICATION_NDJSON).body(result, Sample.class)
				.switchIfEmpty(ServerResponse.status(HttpStatus.CONFLICT).build())
				.onErrorResume(e->ServerResponse.badRequest().bodyValue(e.getMessage()));
	}
}

File: modeler/src/main/kotlin/net/sayaya/document/modeler/sample/SampleRouter.kt
package net.sayaya.document.modeler.sample

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.MediaType.APPLICATION_JSON
import org.springframework.http.codec.ServerSentEvent
import org.springframework.http.codec.multipart.FilePart
import org.springframework.http.codec.multipart.Part
import org.springframework.util.MultiValueMap
import org.springframework.web.reactive.function.BodyInserters
import org.springframework.web.reactive.function.server.*
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Configuration
class SampleRouter(private val handler: SampleHandler) {
    @Bean
    fun documentRouterInstance(): RouterFunction<ServerResponse?> {
        return RouterFunctions
            .route(RequestPredicates.GET("/models/{model}/samples"), this::findSamples)
            .andRoute(RequestPredicates.POST("/models/{model}/samples"), this::uploadSamples)
            .andRoute(RequestPredicates.DELETE("/models/{model}/samples/{id}"), this::removeSamples)
            .andRoute(RequestPredicates.GET("/models/{model}/samples/changes"), this::subscribeSample)
    }

    private fun findSamples(request: ServerRequest): Mono<ServerResponse?> {
        val model = request.pathVariable("model")
        return handler.list(model).collectList()
            .flatMap(ServerResponse.ok().contentType(APPLICATION_JSON)::bodyValue)
            .switchIfEmpty(ServerResponse.noContent().build())
            .onErrorResume (ServerResponse.badRequest()::bodyValue)
    }

    private fun uploadSamples(request: ServerRequest): Mono<ServerResponse?> {
        return handler.upload(request.pathVariable("model"),
            request.multipartData()
                .map { i -> i["files"] }
                .flatMapMany { Flux.fromIterable(it) }
                .cast(FilePart::class.java))
            .collectList()
            .flatMap { ServerResponse.ok().build() }
            .switchIfEmpty(ServerResponse.status(HttpStatus.CONFLICT).build())
            .onErrorResume(ServerResponse.badRequest()::bodyValue)
    }

    private fun removeSamples(request: ServerRequest): Mono<ServerResponse?> {
        val model = request.pathVariable("model")
        val id = request.pathVariable("id")
        return handler.remove(model, id)
            .flatMap { ServerResponse.ok().build() }
            .onErrorResume(ServerResponse.badRequest()::bodyValue)
    }

    private fun subscribeSample(request: ServerRequest): Mono<ServerResponse?> {
        val model = request.pathVariable("model")
        return ServerResponse.ok().contentType(MediaType.TEXT_EVENT_STREAM)
            .body(BodyInserters.fromServerSentEvents(handler.subscribe(model)
                .map { msg -> ServerSentEvent.builder(msg.data).event(msg.type.name).id(msg.data.name()).build() }))
    }
}